pkg/mcs/resourcemanager/server/keyspace_manager.go

```diff
@@ -0,0 +1,451 @@
+		return err
+	}
+	if err := group.persistStates(krgm.keyspaceID, krgm.getStorage()); err != nil {
+		return err
+	}
+	krgm.groups[group.Name] = group
+	return nil
+}
+
+func (krgm *keyspaceResourceGroupManager) modifyResourceGroup(group *rmpb.ResourceGroup) error {
+	if group == nil || group.Name == "" {
+		return errs.ErrInvalidGroup
+	}
+	krgm.Lock()
+	curGroup, ok := krgm.groups[group.Name]
+	krgm.Unlock()
+	if !ok {
+		return errs.ErrResourceGroupNotExists.FastGenByArgs(group.Name)
+	}
+
+	err := curGroup.PatchSettings(group)
+	if err != nil {
+		return err
+	}
+	return curGroup.persistSettings(krgm.keyspaceID, krgm.getStorage())
+}
+
+func (krgm *keyspaceResourceGroupManager) deleteResourceGroup(name string) error {
+	if name == reservedDefaultGroupName {
+		return errs.ErrDeleteReservedGroup
+	}
+	if err := krgm.getStorage().DeleteResourceGroupSetting(krgm.keyspaceID, name); err != nil {
+		return err
+	}
+	krgm.Lock()
+	delete(krgm.groups, name)
+	krgm.Unlock()
+	return nil
+}
+
+func (krgm *keyspaceResourceGroupManager) getResourceGroup(name string, withStats bool) *ResourceGroup {
+	krgm.RLock()
+	defer krgm.RUnlock()
+	if group, ok := krgm.groups[name]; ok {
+		return group.Clone(withStats)
+	}
+	return nil
+}
+
+func (krgm *keyspaceResourceGroupManager) getMutableResourceGroup(name string) *ResourceGroup {
+	krgm.Lock()
+	defer krgm.Unlock()
+	return krgm.groups[name]
+}
+
+func (krgm *keyspaceResourceGroupManager) getResourceGroupList(withStats, includeDefault bool) []*ResourceGroup {
+	krgm.RLock()
+	res := make([]*ResourceGroup, 0, len(krgm.groups))
+	for _, group := range krgm.groups {
+		if !includeDefault && group.Name == reservedDefaultGroupName {
+			continue
+		}
+		res = append(res, group.Clone(withStats))
+	}
+	krgm.RUnlock()
+	sort.Slice(res, func(i, j int) bool {
+		return res[i].Name < res[j].Name
+	})
+	return res
+}
+
+func (krgm *keyspaceResourceGroupManager) getResourceGroupNames(includeDefault bool) []string {
+	krgm.RLock()
+	defer krgm.RUnlock()
+	res := make([]string, 0, len(krgm.groups))
+	for name := range krgm.groups {
+		if !includeDefault && name == reservedDefaultGroupName {
+			continue
+		}
+		res = append(res, name)
+	}
+	return res
+}
+
+func (krgm *keyspaceResourceGroupManager) persistResourceGroupRunningState() {
+	krgm.RLock()
+	keys := make([]string, 0, len(krgm.groups))
+	for k := range krgm.groups {
+		keys = append(keys, k)
+	}
+	krgm.RUnlock()
+	for idx := range keys {
+		krgm.RLock()
+		group, ok := krgm.groups[keys[idx]]
+		if ok {
+			if err := group.persistStates(krgm.keyspaceID, krgm.getStorage()); err != nil {
+				log.Error("persist keyspace resource group state failed",
+					zap.Uint32("keyspace-id", krgm.keyspaceID),
+					zap.String("group-name", group.Name),
+					zap.Int("index", idx),
+					zap.Error(err))
+			}
+		}
+		krgm.RUnlock()
+	}
+}
+
+// TODO: add keyspace name lable to the metrics.
+func (krgm *keyspaceResourceGroupManager) backgroundMetricsFlush(ctx context.Context) {
```

Comment:
Is it better to share a single background routine for the backgroundMetricsFlush for all keyspaces? if there are thousands of keyspaces while each only contains few resource groups, to many background routines may have a bad effect on performance.